source/s3_request_messages.c

```diff
@@ -835,6 +836,49 @@ struct aws_input_stream *aws_s3_message_util_assign_body(
     return NULL;
 }
 
+/* Given all possible ways to send a request body, always return an async-stream */
+struct aws_async_stream *aws_s3_message_util_acquire_async_body_stream(
+    struct aws_allocator *allocator,
+    struct aws_http_message *message,
+    struct aws_byte_cursor send_filepath,
+    struct aws_async_stream *send_async_stream) {
+
+    AWS_PRECONDITION(message);
+
+    /* If user provides async-stream, use it */
+    if (send_async_stream != NULL) {
+        return aws_async_stream_acquire(send_async_stream);
+    }
+
+    /* If user provides filepath, create aws_input_stream to read it, and wrap that in an async-stream */
+    if (send_filepath.len > 0) {
+        struct aws_string *filepath_str = aws_string_new_from_cursor(allocator, &send_filepath);
+        struct aws_input_stream *body_stream =
+            aws_input_stream_new_from_file(allocator, aws_string_c_str(filepath_str));
+        aws_string_destroy(filepath_str);
+        if (body_stream == NULL) {
+            return NULL;
+        }
+        send_async_stream = aws_async_stream_new_from_input_stream(allocator, body_stream);
+        aws_input_stream_release(body_stream);
+        return send_async_stream;
+    }
+
+    /* If user provides HTTP message with aws_input_stream body, wrap that in an async stream */
+    struct aws_input_stream *request_body = aws_http_message_get_body_stream(message);
+    if (request_body) {
+        return aws_async_stream_new_from_input_stream(allocator, request_body);
+    }
+
+    /* Otherwise, no body provided, just create empty async-stream */
+    struct aws_byte_cursor empty_cursor = {.len = 0};
```

Comment:
nit: why not just {0}? to avoid ptr having garbage?

Reply:
If you use designated initializers, it zeroes out anything that you don't explicitly set, which is why I didn't set `.ptr = 0`

I'm pretty sure some compilers warn about `= {0}`. I'll give it a test, but if it doesn't work I'll use good old AWS_ZERO_STRUC(empty_cursor);

Reply:
holy crap it worked! beautiful!

Reply:
nice. i was misremembering the standard on this one - setting any element in the struct does set other elements to 0. {0} initializer is standard since c89 and probably earlier so we should be fine using it everywhere